hazelcast/src/main/java/com/hazelcast/cache/impl/PreJoinCacheConfig.java

```diff
@@ -69,6 +70,40 @@ protected void readKeyValueTypes(ObjectDataInput in)
         setValueClassName(in.readUTF());
     }
 
+    @Override
+    protected void writeFactories(ObjectDataOutput out) throws IOException {
+        SerializationService serializationService = out.getSerializationService();
+        out.writeData(lazyCacheLoaderFactory.getSerializedValue(serializationService));
```

Comment:
I'm guessing previously the EE code serialised factories with :
```
out.writeObject(cacheLoaderFactory);
        out.writeObject(cacheWriterFactory);
        out.writeObject(expiryPolicyFactory);
```
This might be a problem. As I've detected in `com.hazelcast.enterprise.wan.operation.EWRPutOperation#writeInternal`, there is a difference in format between that and:
```
out.writeData(lazyCacheLoaderFactory.getSerializedValue(serializationService));
        out.writeData(cacheWriterFactory.getSerializedValue(serializationService));
        out.writeData(expiryPolicyFactory.getSerializedValue(serializationService));
```
The difference being that the serialization service additionally writes the partition hash. So some 3.9 member might now start getting this data with additionally the partition hash.
Actually I'm not 100% sure here so it may be best if you confirm this.

Reply:
Good one, indeed this serialization block needs to be guarded with version checks to ensure compatibility during RU from 3.9.